### ISSUE TITLE
TypeVar with type_check_only bound

### ISSUE THREAD

**Issue by kltws** [microsoft/pyright#6672]
Currently a class marked as `type_check_only` is flagged as unusable as a `TypeVar` bound.
This particular restriction isn't explicitly mentioned in the documentation, but I would expect this to be allowed as long as the bound is in quotes (or this was in a stub file where they wouldn't be necessary). Consider allowing this as a feature request if not a bug.

```python
from typing import TYPE_CHECKING, TypeVar

if TYPE_CHECKING:
    from typing import Protocol, type_check_only

    @type_check_only
    class _TypeCheckOnly(Protocol):
        ...

# "_TypeCheckOnly" is marked as @type_check_only and can be used only in type annotations
_T = TypeVar('_T', bound='_TypeCheckOnly')

def accepts_t(t: _T) -> _T:
    ...

# "_TypeCheckOnly" is marked as @type_check_only and can be used only in type annotations
def accepts_t_python_12[T: '_TypeCheckOnly'](t: T) -> T:
    ...
```

**Comment by erictraut** [comment]
Hmm, I'm not convinced that pyright's behavior here is incorrect. PEP 484 indicates that:

> @type_check_only, a decorator only available during type checking for use in stub files (see above); marks a class or function as unavailable during runtime

You're using `@type_check_only` in a way that was not intended by the authors of PEP 484. Why are you defining this class under a `TYPE_CHECKING` conditional? I recommend avoiding `TYPE_CHECKING` conditional logic wherever possible because it's effectively "lying" to the type checker and often leads to code that is less robust. Is there a reason why you don't simply declare this protocol class unconditionally in your code?

**Comment by kltws** [comment]
The class defined using `TypedDict` itself is being introspected in this case (`type[_T]` in `__init__`). Instances of `_T` (or rather any dict that conforms to `_T`'s structure) are checked later on, without relying on TypedDict's implementation details on the instances themselves.

```python
_T = TypeVar('_T', bound='class defined with TypedDict')

class TypedDictValidator(Generic[_T]):
    def __init__(self, typed_dict_cls: type[_T]) -> None:
        for key in (typed_dict_cls.__required_keys__ | typed_dict_cls.__optional_keys__):
            ...
    
    def validate(self, instance: _T) -> bool: ...
```

**Comment by erictraut** [comment]
Hmm, that strikes me as an ill-advised use of `TypedDict`. You're definitely in a "gray area" of the type system, and I think you're stretching the type system beyond its intended use here.

As I said above, `TypedDict` is a special form used to define structural types. It's similar to `Protocol`, and I don't think you'd attempt to use `Protocol` in this same manner.

From a type theory standpoint, it doesn't make sense to allow an upper bound of `TypedDict`. I think there's an argument to be made that `type[TypedDict]` could be used as an upper bound. I think that would accommodate your use case.

If you think this use case is important, I recommend starting a thread in the [Python typing discussion forum](https://discuss.python.org/c/typing/32). The recently-adopted [PEP 729](https://peps.python.org/pep-0729/) provides a way for issues like this to be clarified and formalized in the Python typing specification. The first step of this process is to raise the issue and get feedback from the broader community.

If there is an official clarification in the typing spec, pyright will adopt it. In the meantime, I'm going to leave pyright's behavior unchanged.

**Comment by kltws** [comment]
You're probably right that it's stretching the type system. Allowing `[_T: type[TypedDict]]` would allow `TypedDictValidator.__init__` to be typed checked, but the `validate` method would need to be typed as accepting 'instances of _T' which I don't think is possible. I'll settle with the following workaround and consider the matter settled for now. Thank you for help!

```python
_T = TypeVar('_T', bound='Mapping[str, Any]')

class TypedDictValidator(Generic[_T]):
    def __init__(self, typed_dict_cls: type[_T]) -> None:
        required_keys = getattr(typed_dict_cls, '__required_keys__', ())
        ...
    
    def validate(self, instance: _T) -> bool: ...
```

**Comment by XieJiSS** [comment]
Then this would be impossible:

```python
D = TypeVar("D", bound=_TypedDict)
def partial_init(typed_dict: Type[D], **kw: Unpack[D]) -> D:
    ...
```

And specifying `bound=Mapping[str, Any]` is not useful because `Unpack` insists that `D` should be a `TypedDict`.
